src/protocol_auth.c

```diff
@@ -189,7 +193,7 @@ static bool receive_invitation_sptps(void *handle, uint8_t type, const void *dat
 	}
 
 	if(type != 0 || len != 18 || c->status.invitation_used) {
-		return false;
+		return notify_invitation_error(c, "bad request");
```

Comment:
This changes the return value from `false` to `true`, meaning the connection will stay open. We want it to be closed right after sending the error message.

Reply:
How about using a separate status flag in `connection_t`? I don't think there's a way to flush the output buffer *right now* without blocking the thread for an indeterminate amount of time (which is why connection termination was left to the client). If we return `false` here, the connection is terminated without sending anything.

Reply:
Ah yes. The problem is that an attacker might try to join using a false invitation. We don't want them to tie up resources unnecessarily, and we also don't want to give them information that might be used to determine if parts of the join request are valid. So I'm thinking we should not notify tinc join about errors at all. If the connection is closed prematurely, that's the sign that the invitation was not accepted.

Reply:
Makes sense. I removed tarpit for both phases of the connection (pre & post-sptps) since (as you say) the only thing it does it tie up server resources and confuse users (re #255). Calling tarpit on connection burst seems to take care of script kiddies with dictionary bruteforce without worsening experience for legitimate users.

Thanks for spending time on this, I know you could have added this yourself in about 5 seconds; review takes a lot more effort.

Reply:
See my comment about tarpitting in the code. I might have done this in 5 seconds, if I've had the idea of improving this. But you are doing a lot of work now that I might not even have attempted, and doing things different that what I would have done sometimes gives me a new perspective. So I'm fine with these reviews taking more effort, it's a net positive IMO :)